src/nvim/statusline.c

```diff
@@ -1635,50 +1635,43 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, OptIndex op
       if (stcp == NULL) {
         break;
       }
-      bool fold = opt == STL_FOLDCOL;
-      int fdc = fold ? compute_foldcolumn(wp, 0) : 0;
-      int width = fold ? fdc > 0 : wp->w_scwidth;
+      int fdc = opt == STL_FOLDCOL ? compute_foldcolumn(wp, 0) : 0;
+      int width = opt == STL_FOLDCOL ? fdc > 0 : wp->w_scwidth;
 
       if (width <= 0) {
         break;
       }
       foldsignitem = curitem;
 
-      char *p = NULL;
-      if (fold) {
-        schar_T fold_buf[10];
+      if (fdc > 0) {
+        schar_T fold_buf[9];
         fill_foldcolumn(wp, stcp->foldinfo, (linenr_T)get_vim_var_nr(VV_LNUM),
                         0, fdc, NULL, fold_buf);
         stl_items[curitem].minwid = -((stcp->use_cul ? HLF_CLF : HLF_FC) + 1);
         size_t buflen = 0;
         // TODO(bfredl): this is very backwards. we must support schar_T
         // being used directly in 'statuscolumn'
         for (int i = 0; i < fdc; i++) {
-          buflen += schar_get(out_p + buflen, fold_buf[i]);
+          buflen += schar_get(buf_tmp + buflen, fold_buf[i]);
         }
-        p = out_p;
       }
 
-      char buf[SIGN_WIDTH * MAX_SCHAR_SIZE];
-      size_t buflen = 0;
-      varnumber_T virtnum = get_vim_var_nr(VV_VIRTNUM);
+      size_t signlen = 0;
       for (int i = 0; i < width; i++) {
-        if (!fold) {
-          SignTextAttrs *sattr = virtnum ? NULL : &stcp->sattrs[i];
-          p = "  ";
-          if (sattr && sattr->text[0]) {
-            describe_sign_text(buf, sattr->text);
-            p = buf;
+        stl_items[curitem].start = out_p + signlen;
+        if (fdc == 0) {
+          SignTextAttrs *sattrs = get_vim_var_nr(VV_VIRTNUM) ? NULL : &stcp->sattrs[i];
+          if (sattrs->text[0]) {
```

Comment:
Why is there no NULL check in the condition below when one branch of this ternary operator is NULL?